src/node.rs

```diff
@@ -185,246 +198,205 @@ impl<Node> InternalNode<Node> {
+	pub fn with_state<R>(&self, closure: impl FnOnce() -> R) -> R
+	where
+		<TFullCallExecutor<T::Block, T::Executor> as CallExecutor<T::Block>>::Error: std::fmt::Debug,
+	{
+		let id = BlockId::Hash(self.client.info().best_hash);
+		let mut offchain_overlay = OffchainOverlayedChanges::disabled();
+		let mut overlay = OverlayedChanges::default();
+		let changes_trie = backend::changes_tries_state_at_block(&id, self.backend.changes_trie_storage()).unwrap();
+		let mut cache =
+			StorageTransactionCache::<T::Block, <TFullBackend<T::Block> as Backend<T::Block>>::State>::default();
+		let mut extensions = self
+			.client
+			.execution_extensions()
+			.extensions(&id, ExecutionContext::BlockConstruction);
+		let state_backend = self
+			.backend
+			.state_at(id.clone())
+			.expect(&format!("State at block {} not found", id));
+
+		let mut ext = Ext::new(
+			&mut overlay,
+			&mut offchain_overlay,
+			&mut cache,
+			&state_backend,
+			changes_trie.clone(),
+			Some(&mut extensions),
+		);
+		sp_externalities::set_and_run_with_externalities(&mut ext, closure)
+	}
+
+	/// submit some extrinsic to the node, providing the sending account.
+	pub fn submit_extrinsic(
+		&self,
+		call: impl Into<<T::Runtime as frame_system::Config>::Call>,
+		from: <T::Runtime as frame_system::Config>::AccountId,
+	) -> <T::Block as BlockT>::Hash
+	where
+		<T::Runtime as frame_system::Config>::AccountId: Encode,
+		<T::Runtime as frame_system::Config>::Call: Encode,
+		<T::Block as BlockT>::Extrinsic: From<
+			UncheckedExtrinsic<
+				MultiAddress<
+					<T::Runtime as frame_system::Config>::AccountId,
+					<T::Runtime as frame_system::Config>::Index,
+				>,
+				<T::Runtime as frame_system::Config>::Call,
+				MultiSignature,
+				T::SignedExtras,
+			>,
+		>,
+	{
+		let extra = self.with_state(|| T::signed_extras(from.clone()));
+		let signed_data = Some((from.into(), MultiSignature::Sr25519(Default::default()), extra));
+		let ext = UncheckedExtrinsic::<
+			MultiAddress<
+				<T::Runtime as frame_system::Config>::AccountId,
+				<T::Runtime as frame_system::Config>::Index,
+			>,
+			<T::Runtime as frame_system::Config>::Call,
+			MultiSignature,
+			T::SignedExtras,
+		>::new(call.into(), signed_data)
+		.expect("UncheckedExtrinsic::new() always returns Some");
+		let at = self.client.info().best_hash;
+
+		self.compat_runtime()
+			.borrow_mut()
+			.block_on_std(
+				self.pool.submit_one(&BlockId::Hash(at), TransactionSource::Local, ext.into()),
+			)
+			.unwrap()
+	}
+
+	/// Get the events of the most recently produced block
+	pub fn events(&self) -> Vec<EventRecord<T::Runtime>> {
+		self.with_state(|| frame_system::Module::<T::Runtime>::events())
+	}
+
+	/// Checks the node logs for a specific entry.
+	pub fn assert_log_line(&self, content: &str) {
+		futures::executor::block_on(async {
+			use futures::StreamExt;
+
+			while let Some(log_line) = self.log_stream.borrow_mut().next().await {
```

Comment:
Why do we use `RefCell` instead of simply doing a `&mut self` API? Given that the struct is not really thread-safe, we risk panics in case we borrow fields that are already mutably borrowed somewhere higher in the stack.

Reply:
> we risk panics in case we borrow fields that are already mutably borrowed somewhere higher in the stack.

so like you said, the struct isn't thread-safe, so it's impossible that this function would be called while a previous call to this function hasn't terminated.

Reply:
Then it's another argument to introduce `&mut self` API and rely on borrow checker instead of human review.
Imagine this:
```rust
fn utility_function(&self) -> u32 {
  self.some_parameter.borrow_mut().get_and_cache_value()
}

fn another_function(&self) -> String {
   let some_parameter = self.some_parameter.borrow_mut();
   let value = self.utility_function();
   do_something_else(value, some_parameter)
}
```

This code is obviously incorrect even in single-threaded and it isn't immediately obvious that it's the case. Using `RefCell` is the main culrpit of introducing this kind of bugs in the code. It would work just fine if both functions were simply `&mut self` - the compiler would ensure that you do correct things here.

Reply:
ahh, I see what you mean, I felt &mut self makes the api unergonomic but apparently, the tradeoff is correctness.lemme revert